actors/builtin/miner/sector_map.go

```diff
@@ -0,0 +1,151 @@
+package miner
+
+import (
+	"math"
+	"sort"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-bitfield"
+)
+
+// Maps deadlines to partition maps.
+type DeadlineSectorMap map[uint64]PartitionSectorMap
+
+// Maps partitions to sector bitfields.
+type PartitionSectorMap map[uint64]*bitfield.BitField
+
+// Check validates all bitfields and counts the number of partitions & sectors
+// contained within the map, and returns an error if they exceed the given
+// maximums.
+func (dm DeadlineSectorMap) Check(maxPartitions, maxSectors uint64) error {
+	partitionCount, sectorCount, err := dm.Count()
+	if err != nil {
+		return xerrors.Errorf("failed to count sectors: %w", err)
+	}
+	if partitionCount > maxPartitions {
+		return xerrors.Errorf("too many partitions %d, max %d", partitionCount, maxPartitions)
+	}
+
+	if sectorCount > maxSectors {
+		return xerrors.Errorf("too many sectors %d, max %d", sectorCount, maxSectors)
+	}
+
+	return nil
+}
+
+// Count counts the number of partitions & sectors within the map.
+func (dm DeadlineSectorMap) Count() (partitions, sectors uint64, err error) {
+	for dlIdx, pm := range dm { //nolint:nomaprange
+		partCount, sectorCount, err := pm.Count()
+		if err != nil {
+			return 0, 0, xerrors.Errorf("when counting deadline %d: %w", dlIdx, err)
+		}
+		if partCount > math.MaxUint64-partitions {
```

Comment:
Go doesn't have overflow-safe arithmetic functions?

It's a great insight to add these checks. We should write our own arithmetic wrappers and do this everywhere.

Reply:
Unfortunately, no, they don't. It's not an issue in 99% of cases, but this is one of the few where it is (because someone can pass an arbitrarily large bitfield).

Of course, now that we're restricting bitfield runs to `2**63-1` instead of `2**64`, this may become a non-issue.